src/language/pattern-types.md

```diff
@@ -0,0 +1,406 @@
+```
+
+## Logical-and	
+
+`subpattern1 && subpattern2`
+
+A pair of patterns separated by `&&` matches only if both subpatterns match. If the
+left branch does not match, the right branch is not evaluated.
+
+Subpatterns in a logical-and pattern can bind variables, but the variables in
+each subpattern must not overlap, because they will both be bound if the pattern
+matches:
+
+```dart
+switch ((1, 2)) {
+  // Error, both subpatterns attempt to bind 'b'.
+  case (var a, var b) && (var b, var c): // ... 
+}
+```
+
+## Logical-or
+
+`subpattern1 || subpattern2`
+
+A logical-or pattern separates subpatterns by `||` and matches if any of the
+branches match. Branches are evaluated left-to-right. Once a branch matches, the
+rest are not evaluated.
+
+```dart
+var isPrimary = switch (Color color) {
+  Color.red || Color.yellow || Color.blue => true,
+  _ => false
+};
+```
+
+Subpatterns in a logical-or pattern can bind variables, but the branches must
+define the same set of variables, because only one branch will be evaluated when
+the pattern matches.
+
+{% comment %}
+[TODO: move the below content to the switch page, doesn't belong here]
+
+Logical-or patterns are useful for having multiple cases share a body in switch
+expressions or statements.
+
+```dart
+var isPrimary = switch (color) {
+  Color.red || Color.yellow || Color.blue => true,
+  _ => false
+};
+```
+
+Switch statements can have multiple cases share a body without using logical-or
+patterns, but they are still uniquely useful for allowing multiple cases to share
+a guard: 
+
+```dart
+switch (shape) {
+  case Square(size: var s) || Circle(size: var s) when s > 0:
+    print('Non-empty symmetric shape');
+}
+```
+
+{% endcomment %}
+
+## Map
+
+`{"key": subpattern1, someConst: subpattern2}`
+
+Map patterns match values that implement [`Map`][], and then recursively 
+match its subpatterns against the map’s keys to destructure them.
+
+Map patterns don't require the pattern to match the entire map. A map pattern
+ignores any keys that the map contains that aren't matched by the pattern.
+
+## Null-assert	
+
+`subpattern!`
+
+Null-assert patterns match first if the object is not null, then on the value.
+They permit non-null values to flow through, but [throw][] if the matched value is null. 
+
+```dart
+List<String?> row = // ...
+
+// If the first column is 'user', we expect to have a name after it.
+switch (row) {
+  case ['user', var name!]: // ...
+    // 'name' is a non-nullable string here.
+}
+```
+
+Use the null-assert pattern to eliminate `null` values from variable declaration
+patterns.
+
+## Null-check	
+
+`subpattern?`
+
+Null-check patterns match first if the value is not null, and then match the inner
+pattern against that same value.
+
+```dart
+String? maybeString = // ...
+switch (maybeString) {
+  case var s?: ...
+    // 's' has type non-nullable String here.
+}
+```
+
+Use a null-check pattern while matching if you need to ensure `null` values
```

Comment:
I don't understand this sentence.  Doesn't a null-check pattern do exactly that: silently treat a null value as a match failure?

Should this be "Use a null-assert pattern ..."?  Maybe the sentence above, about null-assert patterns, should be the one about null-check patterns, and they are accidentally switched.

Reply:
I do think the other sentence is about null checks - it is "silently eliminating null (matches) from a pattern", without throwing. I don't see how assets could ever be silent.

Reply:
Got it. The null check pattern is about purposefully treating null values as match failures, and null assert is about _not_ treating them silently. I kept the line in both sections, emphasizing the difference